internal/powerplatform/services/data_record/odata.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+package powerplatform
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+func BuildODataQueryFromModel(model *DataRecordListDataSourceModel) (string, map[string]string, error) {
+	var resultQuery = ""
+	var headers = make(map[string]string)
+
+	appendQuery(&resultQuery, buildODataSelectPart(model.Select))
+	appendQuery(&resultQuery, buildODataFilterPart(model.Filter.ValueStringPointer()))
+	appendQuery(&resultQuery, buildOdataApplyPart(model.Apply.ValueStringPointer()))
+	appendQuery(&resultQuery, buildODataOrderByPart(model.OrderBy.ValueStringPointer()))
```

Comment:
there are a lot of functions here for just adding a query part.  I get that you're trying to workaround the "$" in the key getting encoded.  Though a little less efficient I wonder if it would cut down on code to just have one url.Values that you add to using 

```golang
v := url.Values{}
if(model.Select != nil) { v.Add("$select", url.QueryEscape(model.Select)) }
...

queryString := url.QueryUnescape(v.Encode())

```

Reply:
this will just make the code less readable. I don't see a value of using url struct if I have to write bigger, less clean method with it. 
I've also refactored existing code a bit more